### ISSUE TITLE
could not decrypt bytes

### ISSUE THREAD

**Issue by dungnguyen10989** [oblador/react-native-keychain#177]
On iOs or debug android, everything is well, but when I run release on android, I've received this error. How to resolve this ?

**Comment by ninjz** [comment]
Same issue, though I only am getting this issue when having released to the Play store, and not when installed locally.

**Comment by jerryheir** [comment]
Any solution for this please????? I downloaded my app from the playstore and it says the same

**Comment by janpieterz** [comment]
We're running into this on only one phone through a support request, mimicking API level etc in a simulator has not resulted in anything.

**Comment by RossKinsella** [comment]
I'm seeing this happen in one of my apps
1k events for 330 unique users. I have 1k unique users
I have events from android 9, 8 & 6
I am using it to store multiple keys

Not clear to me how to resolve this apart from removing this library. Any thoughts?

**Comment by RossKinsella** [comment]
I'm seeing this happen in one of my apps
My error reports are telling me 33% of my user base have encountered this issue

I have events from android 9, 8 & 6
I am using it to store multiple keys

Not clear to me how to resolve this apart from removing this library. Any thoughts?

**Comment by janpieterz** [comment]
Is someone in this thread able to replicate on a simulator or device? As the exception gets thrown with more data, it might offer more insights into the issue. 

If we have a reliable configuration reproducing on a simulator I'm happy to put in the effort to dig deeper.

**Comment by astapinski** [comment]
We were having this in our production app appearing in Bugsnag, and I think I've tracked it down. For us, this was happening when a user was backing up and restoring their phone on a new one (i.e. Keystore doesn't come across) but our app data was. So our app was asking for a key that effectively didn't exist and we were getting this error. I'm making a PR for a better message as well as a crash on Android 4.x

TL;DR Catch a promise rejection when calling getPassword() as this error may be around the key not existing. Read on for more.

Could not decrypt bytes is coming around in our case because the key being passed in was null. This resulted in an error in the line:

`cipher.init(Cipher.DECRYPT_MODE, key, ivParams);`

that _java.security.InvalidKeyException: Only SecretKey is supported_. This error is hidden because the error catch for _Could not decrypt bytes_ doesn't append the message, and React Native doesn't expose the inner exceptions (at least not to Bugsnag's reporting?).

So I made a PR to make a better error message, but really the fix was to catch a Promise rejection and handle it:

A sample of this from our app's code

```
return Keychain.getGenericPassword(keychain).then((credentials) => {
        // regular code here
  }).catch(ex => {
    console.error(`An error occurred getting the DB encryption key from the keychain: ${ex}`);
    return;
  });
```

PR is here: https://github.com/oblador/react-native-keychain/pull/209

**Comment by janpieterz** [comment]
Interesting, I get this error on a piece of code that certainly doesn't call getGenericPassword first, and it's only surfacing on one user's phone. Curious to see if the PR once landed will surface a different error for this user.

**Comment by astapinski** [comment]
At the least, the PR would expose the inner exception's error message by concatenating it on to "Could not decrypt bytes". There's a block of code there that could go wrong so yeah, you should get more info. No idea when the PR will land though, I responded to a comment and that's the last state of it.

**Comment by fbartho** [comment]
We've run into this in the wild as well. We're not however supporting Android 4.x devices, so I wonder if this is a different issue. Our lowest device support is set to Android 5.1.1 (Lollipop)

**Comment by astapinski** [comment]
If you look at the code that results in this error, "Could not decrypt bytes" is in a try/catch block that catches any exception. So yeah there's a bunch of ways it could likely come up, that's why in the PR I made the actual exception message will get concatenated on. I haven't seen any change in its status though :|

**Comment by astapinski** [comment]
@fbartho Sorry, one thing I should clarify is the crash on Android 4.x is right at app startup. The could not decrypt bytes error we were getting in our app is simply the key wasn't present (ours was a logic error with our app and backup/restore of a phone) and we had to handle the promise rejection.

**Comment by astapinski** [comment]
The PR is on master now, so you may want to try it out and see if it changes the error (it should, at least you would get the inner exception message). YMMV, so you may want to wait for an official release/verification for a production release. We're going ahead with it though for our company's app.

**Comment by denchp** [comment]
I'm running into this on a Pixel 2, Android 9...

Is there a work around or is this essentially, 'wait for them to release a fix'?

**Comment by TomMahle** [comment]
Not sure if this will be helpful, but I was able to grab a stack trace for this issue from logcat:

06-04 15:07:28.320 2812-3525/com.mypackage E/RNKeychainManager: Unknown error: Could not decrypt bytes: ShortBufferException is not expected
06-04 15:07:28.356 2812-3524/com.mypackage E/ReactNativeJS: { [Error: Unknown error: Could not decrypt bytes: ShortBufferException is not expected]
      framesToPop: 1,
      nativeStackAndroid: 
       [ { methodName: 'decrypt',
           lineNumber: 184,
           file: 'CipherStorageKeystoreAESCBC.java' },
         { methodName: 'decryptCredentials',
           lineNumber: 134,
           file: 'KeychainModule.java' },
         { methodName: 'getGenericPasswordForOptions',
           lineNumber: 113,
           file: 'KeychainModule.java' },
         { methodName: 'getInternetCredentialsForServer',
           lineNumber: 206,
           file: 'KeychainModule.java' },
         { methodName: 'invoke', lineNumber: -2, file: 'Method.java' },
         { methodName: 'invoke',
           lineNumber: 372,
           file: 'JavaMethodWrapper.java' },
         { methodName: 'invoke',
           lineNumber: 158,
           file: 'JavaModuleWrapper.java' },
         { methodName: 'run',
           lineNumber: -2,
           file: 'NativeRunnable.java' },
         { methodName: 'handleCallback',
           lineNumber: 873,
           file: 'Handler.java' },
         { methodName: 'dispatchMessage',
           lineNumber: 99,
           file: 'Handler.java' } ],
      userInfo: null,
      code: 'E_CRYPTO_FAILED',
      line: 2106,
      column: 26,
      sourceURL: 'http://localhost:8081/index.android.delta?platform=android&dev=true&minify=false' }

**Comment by TomMahle** [comment]
Might be a separate issue, actually. Investigating further.

**Comment by mandrigin** [comment]
I don't have a Pixel 3 nearby to test, but what if you increase the buffer size in this line: https://github.com/oblador/react-native-keychain/blob/master/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC.java#L234? will that help?`

**Comment by raimonkh** [comment]
We can see this happening only on android. We have no clue as why it is (yet).
we are on lib version 5
